src/Razor/src/Microsoft.VisualStudio.RazorExtension/Snippets/SnippetService.cs

```diff
@@ -0,0 +1,221 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Editor.Razor;
+using Microsoft.VisualStudio.Editor.Razor.Snippets;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.RazorExtension.Snippets;
+
+internal class SnippetService
+{
+    private readonly JoinableTaskFactory _joinableTaskFactory;
+    private readonly IAsyncServiceProvider _serviceProvider;
+    private readonly SnippetCache _snippetCache;
+    private IVsExpansionManager? _vsExpansionManager;
+
+    private readonly object _cacheGuard = new();
+
+    private static readonly Guid s_CSharpLanguageId = new("694dd9b6-b865-4c5b-ad85-86356e9c88dc");
+    private static readonly Guid s_HtmlLanguageId = new("9bbfd173-9770-47dc-b191-651b7ff493cd");
+
+    private static readonly Dictionary<Guid, ImmutableHashSet<string>> s_ignoredSnippets = new()
+    {
+        {
+            // These are identified as the snippets that are provided by the C# Language Server. The list is found
```

Comment:
Random thought: Should we remove the restriction on the Roslyn side, and let them give us all C# snippets, and then contribute this code, or some form of it, to Web Tools for html snippets?

Reply:
Maybe. I think if we can get the editor to allow access to the snippet parsing code for everyone, then adding them to the html server and csharp server likely would be the right move.